### ISSUE TITLE
Mypy does not report error for incorrect return annotation for NoReturn functions

### ISSUE THREAD

**Issue by gousaiyang** [python/mypy#8841]
The following function never returns, so its return type should be annotated as `NoReturn`. Here the return type is marked as `int`, which is incorrect. But mypy does not report any errors for this.

```python
def foo() -> int:
    raise RuntimeError

foo()
```

**Comment by AlexWaygood** [comment]
I believe this is deliberate: it's quite a common pattern to have this kind of inheritance structure:

```python
class Abstract:
    def method(self) -> int:
        raise NotImplementedError

class Concrete(Abstract):
    def method(self) -> int:
        return 5
```

If mypy reported errors for all functions that unconditionally raised exceptions, this kind of inheritance structure wouldn't be possible.

I suppose you could argue, though, that mypy could distinguish between functions declared in the global scope and methods declared in class bodies.

**Comment by JelleZijlstra** [comment]
Mypy is correct here, all types are compatible with `NoReturn`.
